src/main/java/org/reaktivity/nukleus/http2/internal/stream/Http2ServerFactory.java

```diff
@@ -2187,16 +2245,20 @@ private void onDecodeData(
                 {
                     exchange.cleanup(traceId, authorization);
                     doEncodeRstStream(traceId, authorization, streamId, error);
+                    progress += payloadLength.value;
                 }
                 else
                 {
-                    if (dataLength > 0)
+                    payloadLength.set(payload.capacity());
+
+                    if (payloadLength.value > 0)
                     {
-                        final DirectBuffer payload = http2Data.payload();
-                        exchange.doRequestData(traceId, authorization, payload, 0, payload.capacity());
+                        exchange.doRequestData(traceId, authorization, payload, payloadLength);
+                        progress = payloadLength.value;
                     }
 
-                    if (endRequest)
+                    final int remaining = maxDecodableDataBytes - progress;
+                    if (endRequest && (remaining == 0))
```

Comment:
This reference to `maxDecodableDataBytes` is still not quite working with the right concepts, since `decodableDataBytes` is logically owned by the `decodeXXX` methods.

Consider changing the method signature to something like:
```
        private int onDecodeData(
            long traceId,
            long authorization,
            int streamId,
            byte flags,
            int deferred,
            DirectBuffer payload)
        {
```
Then we have a clear concept of how many bytes are deferred (still to come) after this payload fragment.

When `doRequestData` returns, we might discover that more bytes are deferred due to flow control.
```
                    final int payloadLength = payload.capacity();

                    if (payloadLength.value > 0)
                    {
                        payloadRemaining.set(payloadLength);
	                exchange.doRequestData(traceId, authorization, payload, payloadRemaining);
                        progress = payloadLength - payloadRemaining.value;
                        deferred += payloadRemaining.value;
                    }
```
which requires `doRequestData` to indicate what is left still to be sent, not how much was sent, so if `doRequestData` did nothing, then the value of `payloadRemaining` would remain unchanged.

Finally, the check to determine when to `doRequestEnd(...)` becomes:
```
if (deferred == 0 && Http2Flags.endStream(flags))
{
...
}
```